test/dynamo/test_optimizers.py

```diff
@@ -97,6 +98,42 @@ def setUpClass(cls):
 for opt in optimizers:
     setattr(OptimizerTests, "test_" + opt.__name__.lower(), make_test(opt))
 
+class End2EndTests(torch._dynamo.test_case.TestCase):
+
+    # https://github.com/pytorch/torchdynamo/issues/1604
+    def test_optimizing_over_tensor_with_requires_grad(self):
+        class Net(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+
+            def forward(self, x, y):
+                z = torch.bmm(x, y)
+                z = torch.flatten(z, 1)
+                return z
+
+        def training_iter_fn(batch, model, optimizer):
+            optimizer.zero_grad()
```

Comment:
run line 115 through 120 twice, just to double-test it for cudagraphs / runtime-compilation :)